### DISCUSSION TITLE
REST API Usage Documentation / Examples

### DISCUSSION THREAD

**Discussion by charlesritchea** [argoproj/argo-workflows#13334] [Q&A]
I found the OpenAPI spec https://argo-workflows.readthedocs.io/en/latest/swagger/ but I can't make heads or tails of it. Maybe I'm stupid, but I don't have any idea how to translate this information into something useful like a cUrl command. Should I just focus on a client library and reverse engineer? What am I missing?

**Comment by charlesritchea** [comment]
I guess this is better even if it's only one example. It gives me a vague idea of what to do https://argo-workflows.readthedocs.io/en/latest/rest-examples/

**Reply by agilgur5** [reply]
Yea that example should help with the structure. The first page is pure OpenAPI spec, so it effectively assumes you know how to read & use them. And it lacks much API docs too 😕

You can also take the spec and plop it into Postman/Insomnia/Hopscotch (etc). Those and other tools can generate cURL commands from the spec.

Also there are a few auto-generated SDKs as well (which are not that great) as well as custom SDKs like Hera (which are great). See https://argo-workflows.readthedocs.io/en/latest/client-libraries/

PRs welcome to improve docs!

**Comment by charlesritchea** [comment]
Yeah, I was going to focus on the Java client, which has atrocious class names! but even then this seems to be a common problem with API docs from my experience. I want real world examples of Json bodies, not just a hypertextual list of all possible parameters  and I have to use my imagination to figure out what is actually needed. I guess I'll manage. I'm only exploring Argo as a replacement for a homegrown system at the moment. Which is why API integration is important. I guess I'm just a different kind of neurodivergent from people that love swagger tho 🤷‍♀️

**Reply by agilgur5** [reply]
> I want real world examples of Json bodies, not just a hypertextual list of all possible parameters  and I have to use my imagination to figure out what is actually needed

For sure; that's kind of what happens with generated docs and clients. Custom SDKs need a lot of maintenance (Hera has a whole team behind it, a substantial portion of which are paid by employer to work on it too), so it's non-trivial for every API to have them, especially in the often heavily under-resourced and volunteer-run world of OSS.

The Argo API is also mainly an intermediary to k8s, so most of it inherits the same specs as the k8s CRs (which do have plenty of examples and docs).
There's a couple wrapper options around the CR specs that are poorly documented though. They make more sense if you've used the CLI or UI; they're options that make sense to have for certain use-cases/commands (usually, or are legacy).

> I guess I'm just a different kind of neurodivergent from people that love swagger tho 🤷‍♀️

Idk if anyone "loves swagger" 😅 It's just a standard API interface to follow that helps a heck of a lot with consistency, automated tooling, etc. Codegen ain't always great (and rarely perfect), but it can be a heck of a lot better than nothing or unmaintained, unidiomatic manual writing and makes integrations more straightforward at least (but not necessarily easy still).

Argo has a lot of different codegen and it is not the most fun to use, but it does get the job done at least. In my experience, I think that applies to most generated tooling; "not perfect, but gets the job done" tends to be an expectation that isn't disappointing 😅
